x-pack/legacy/plugins/code/server/plugin.ts

```diff
@@ -0,0 +1,317 @@
+    this.serverOptions = {} as ServerOptions;
+    this.indexScheduler = {} as IndexScheduler;
+    this.updateScheduler = {} as UpdateScheduler;
+    this.lspService = {} as LspService;
+  }
+
+  // TODO: options is not a valid param for the setup() api
+  // of the new platform. Will need to pass through the configs
+  // correctly in the new platform.
+  public setup(core: CoreSetup, options: any) {
+    const { server } = core.http as any;
+
+    this.log = new Logger(server);
+    this.serverOptions = new ServerOptions(options, server.config());
+
+    const xpackMainPlugin: XPackMainPlugin = server.plugins.xpack_main;
+    xpackMainPlugin.registerFeature({
+      id: 'code',
+      name: i18n.translate('xpack.code.featureRegistry.codeFeatureName', {
+        defaultMessage: 'Code',
+      }),
+      icon: 'codeApp',
+      navLinkId: 'code',
+      app: ['code', 'kibana'],
+      catalogue: [], // TODO add catalogue here
+      privileges: {
+        all: {
+          api: ['code_user', 'code_admin'],
+          savedObject: {
+            all: [],
+            read: ['config'],
+          },
+          ui: ['show', 'user', 'admin'],
+        },
+        read: {
+          api: ['code_user'],
+          savedObject: {
+            all: [],
+            read: ['config'],
+          },
+          ui: ['show', 'user'],
+        },
+      },
+    });
+  }
+
+  // TODO: CodeStart will not have the register route api.
+  // Let's make it CoreSetup as the param for now.
+  public async start(core: CoreSetup) {
+    // called after all plugins are set up
+    const { server } = core.http as any;
+
+    const codeNodeUrl = this.serverOptions.codeNodeUrl;
+    const rndString = crypto.randomBytes(20).toString('hex');
+    checkRoute(server, rndString);
+    if (codeNodeUrl) {
+      const checkResult = await this.retryUntilAvailable(
+        async () => await checkCodeNode(codeNodeUrl, this.log, rndString),
+        5000
+      );
+      if (checkResult.me) {
+        this.isCodeNode = true;
+        await this.initCodeNode(core, this.serverOptions);
+      } else {
+        await this.initNonCodeNode(codeNodeUrl, core);
+      }
+    } else {
+      // codeNodeUrl not set, single node mode
+      this.isCodeNode = true;
+      await this.initCodeNode(core, this.serverOptions);
+    }
+  }
+
+  public async stop() {
+    if (this.isCodeNode) {
+      this.gitOps.cleanAllRepo();
+      if (!this.serverOptions.disableIndexScheduler) {
+        this.indexScheduler.stop();
+      }
+      this.updateScheduler.stop();
+      this.queue.destroy();
+      await this.lspService.shutdown();
+    }
+  }
+
+  private async initNonCodeNode(url: string, core: CoreSetup) {
+    const { server } = core.http as any;
+    this.log.info(
+      `Initializing Code plugin as non-code node, redirecting all code requests to ${url}`
+    );
+    redirectRoute(server, url, this.log);
+  }
+
+  private async initCodeNode(core: CoreSetup, serverOptions: ServerOptions) {
```

Comment:
nit: This is quite a bit code to have in your Plugin class. I'd recommend putting this in a different file to make reading this class easy.

Reply:
yes, there is indeed a separate PR refactoring this giant block.